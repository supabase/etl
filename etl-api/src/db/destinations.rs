use etl_config::SerializableSecretString;
use etl_config::shared::DestinationConfig;
use secrecy::ExposeSecret;
use serde::{Deserialize, Serialize};
use sqlx::PgExecutor;
use std::fmt::Debug;
use thiserror::Error;

use crate::db::serde::{
    DbDeserializationError, DbSerializationError, decrypt_and_deserialize_from_value,
    encrypt_and_serialize,
};
use crate::encryption::{
    Decrypt, DecryptionError, Encrypt, EncryptedValue, EncryptionError, EncryptionKey,
    decrypt_text, encrypt_text,
};

impl Encrypt<EncryptedDestinationConfig> for DestinationConfig {
    fn encrypt(
        self,
        encryption_key: &EncryptionKey,
    ) -> Result<EncryptedDestinationConfig, EncryptionError> {
        match self {
            Self::Memory => Ok(EncryptedDestinationConfig::Memory),
            Self::BigQuery {
                project_id,
                dataset_id,
                service_account_key,
                max_staleness_mins,
                max_concurrent_streams,
            } => {
                let encrypted_service_account_key = encrypt_text(
                    service_account_key.expose_secret().to_owned(),
                    encryption_key,
                )?;

                Ok(EncryptedDestinationConfig::BigQuery {
                    project_id,
                    dataset_id,
                    service_account_key: encrypted_service_account_key,
                    max_staleness_mins,
                    max_concurrent_streams,
                })
            }
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EncryptedDestinationConfig {
    Memory,
    BigQuery {
        project_id: String,
        dataset_id: String,
        service_account_key: EncryptedValue,
        #[serde(skip_serializing_if = "Option::is_none")]
        max_staleness_mins: Option<u16>,
        #[serde(skip_serializing_if = "Option::is_none")]
        max_concurrent_streams: Option<usize>,
    },
}

impl Decrypt<DestinationConfig> for EncryptedDestinationConfig {
    fn decrypt(self, encryption_key: &EncryptionKey) -> Result<DestinationConfig, DecryptionError> {
        match self {
            Self::Memory => Ok(DestinationConfig::Memory),
            Self::BigQuery {
                project_id,
                dataset_id,
                service_account_key: encrypted_service_account_key,
                max_staleness_mins,
                max_concurrent_streams,
            } => {
                let service_account_key = SerializableSecretString::from(decrypt_text(
                    encrypted_service_account_key,
                    encryption_key,
                )?);

                Ok(DestinationConfig::BigQuery {
                    project_id,
                    dataset_id,
                    service_account_key,
                    max_staleness_mins,
                    max_concurrent_streams,
                })
            }
        }
    }
}

#[derive(Debug, Error)]
pub enum DestinationsDbError {
    #[error("Error while interacting with Postgres for destinations: {0}")]
    Database(#[from] sqlx::Error),

    #[error("Error while serializing destination config: {0}")]
    DbSerialization(#[from] DbSerializationError),

    #[error("Error while deserializing destination config: {0}")]
    DbDeserialization(#[from] DbDeserializationError),
}

pub struct Destination {
    pub id: i64,
    pub tenant_id: String,
    pub name: String,
    pub config: DestinationConfig,
}

pub async fn create_destination<'c, E>(
    executor: E,
    tenant_id: &str,
    name: &str,
    config: DestinationConfig,
    encryption_key: &EncryptionKey,
) -> Result<i64, DestinationsDbError>
where
    E: PgExecutor<'c>,
{
    let config = encrypt_and_serialize(config, encryption_key)?;

    let record = sqlx::query!(
        r#"
        insert into app.destinations (tenant_id, name, config)
        values ($1, $2, $3)
        returning id
        "#,
        tenant_id,
        name,
        config
    )
    .fetch_one(executor)
    .await?;

    Ok(record.id)
}

pub async fn read_destination<'c, E>(
    executor: E,
    tenant_id: &str,
    destination_id: i64,
    encryption_key: &EncryptionKey,
) -> Result<Option<Destination>, DestinationsDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        select id, tenant_id, name, config
        from app.destinations
        where tenant_id = $1 and id = $2
        "#,
        tenant_id,
        destination_id,
    )
    .fetch_optional(executor)
    .await?;

    let destination = match record {
        Some(record) => {
            let config = decrypt_and_deserialize_from_value::<
                EncryptedDestinationConfig,
                DestinationConfig,
            >(record.config, encryption_key)?;

            let destination = Destination {
                id: record.id,
                tenant_id: record.tenant_id,
                name: record.name,
                config,
            };

            Some(destination)
        }
        None => None,
    };

    Ok(destination)
}

pub async fn update_destination<'c, E>(
    executor: E,
    tenant_id: &str,
    name: &str,
    destination_id: i64,
    config: DestinationConfig,
    encryption_key: &EncryptionKey,
) -> Result<Option<i64>, DestinationsDbError>
where
    E: PgExecutor<'c>,
{
    let config = encrypt_and_serialize(config, encryption_key)?;

    let record = sqlx::query!(
        r#"
        update app.destinations
        set config = $1, name = $2
        where tenant_id = $3 and id = $4
        returning id
        "#,
        config,
        name,
        tenant_id,
        destination_id
    )
    .fetch_optional(executor)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn delete_destination<'c, E>(
    executor: E,
    tenant_id: &str,
    destination_id: i64,
) -> Result<Option<i64>, DestinationsDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        delete from app.destinations
        where tenant_id = $1 and id = $2
        returning id
        "#,
        tenant_id,
        destination_id
    )
    .fetch_optional(executor)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn read_all_destinations<'c, E>(
    executor: E,
    tenant_id: &str,
    encryption_key: &EncryptionKey,
) -> Result<Vec<Destination>, DestinationsDbError>
where
    E: PgExecutor<'c>,
{
    let records = sqlx::query!(
        r#"
        select id, tenant_id, name, config
        from app.destinations
        where tenant_id = $1
        "#,
        tenant_id,
    )
    .fetch_all(executor)
    .await?;

    let mut destinations = Vec::with_capacity(records.len());
    for record in records {
        let config = decrypt_and_deserialize_from_value::<
            EncryptedDestinationConfig,
            DestinationConfig,
        >(record.config.clone(), encryption_key)?;

        let destination = Destination {
            id: record.id,
            tenant_id: record.tenant_id,
            name: record.name,
            config,
        };
        destinations.push(destination);
    }

    Ok(destinations)
}

pub async fn destination_exists<'c, E>(
    executor: E,
    tenant_id: &str,
    destination_id: i64,
) -> Result<bool, DestinationsDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        select exists (select id
        from app.destinations
        where tenant_id = $1 and id = $2) as "exists!"
        "#,
        tenant_id,
        destination_id,
    )
    .fetch_one(executor)
    .await?;

    Ok(record.exists)
}

#[cfg(test)]
mod tests {
    use aws_lc_rs::aead::RandomizedNonceKey;
    use etl_config::SerializableSecretString;
    use etl_config::shared::DestinationConfig;

    use crate::db::destinations::EncryptedDestinationConfig;
    use crate::db::serde::{decrypt_and_deserialize_from_value, encrypt_and_serialize};
    use crate::encryption::EncryptionKey;

    #[test]
    pub fn destination_config_json_deserialization() {
        let config = r#"{
            "big_query": {
                "project_id": "project-id",
                "dataset_id": "dataset-id",
                "service_account_key": "service-account-key",
                "max_staleness_mins": 42
            }
        }"#;

        let deserialized = serde_json::from_str::<DestinationConfig>(config);
        insta::assert_debug_snapshot!(deserialized.unwrap());
    }

    #[test]
    pub fn destination_config_json_serialization() {
        let config = DestinationConfig::BigQuery {
            project_id: "project-id".to_string(),
            dataset_id: "dataset-id".to_string(),
            service_account_key: SerializableSecretString::from("service-account-key".to_string()),
            max_staleness_mins: Some(42),
            max_concurrent_streams: None,
        };

        insta::assert_json_snapshot!(config);
    }

    #[test]
    pub fn destination_config_json_encryption() {
        let key_bytes = [42u8; 32];
        let key = RandomizedNonceKey::new(&aws_lc_rs::aead::AES_256_GCM, &key_bytes).unwrap();
        let encryption_key = EncryptionKey { id: 1, key };

        let config = DestinationConfig::BigQuery {
            project_id: "project-id".to_string(),
            dataset_id: "dataset-id".to_string(),
            service_account_key: SerializableSecretString::from("supersecretkey".to_string()),
            max_staleness_mins: Some(99),
            max_concurrent_streams: None,
        };

        let config_in_db = encrypt_and_serialize::<DestinationConfig, EncryptedDestinationConfig>(
            config.clone(),
            &encryption_key,
        )
        .unwrap();
        insta::assert_json_snapshot!(config_in_db, {
            ".big_query.service_account_key" => "[key]"
        });

        let deserialized_config = decrypt_and_deserialize_from_value::<
            EncryptedDestinationConfig,
            DestinationConfig,
        >(config_in_db, &encryption_key)
        .unwrap();
        insta::assert_debug_snapshot!(deserialized_config);
    }
}
