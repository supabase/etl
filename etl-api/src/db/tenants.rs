use sqlx::PgExecutor;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum TenantsDbError {
    #[error("Error while interacting with Postgres for tenants: {0}")]
    Database(#[from] sqlx::Error),
}

pub struct Tenant {
    pub id: String,
    pub name: String,
}

pub async fn create_tenant<'c, E>(
    executor: E,
    tenant_id: &str,
    tenant_name: &str,
) -> Result<String, TenantsDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        insert into app.tenants (id, name)
        values ($1, $2)
        returning id
        "#,
        tenant_id,
        tenant_name,
    )
    .fetch_one(executor)
    .await?;

    Ok(record.id)
}

pub async fn create_or_update_tenant<'c, E>(
    executor: E,
    tenant_id: &str,
    tenant_name: &str,
) -> Result<String, TenantsDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        insert into app.tenants (id, name)
        values ($1, $2)
        on conflict (id) do update set name = $2
        returning id
        "#,
        tenant_id,
        tenant_name,
    )
    .fetch_one(executor)
    .await?;

    Ok(record.id)
}

pub async fn read_tenant<'c, E>(
    executor: E,
    tenant_id: &str,
) -> Result<Option<Tenant>, TenantsDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        select id, name
        from app.tenants
        where id = $1
        "#,
        tenant_id
    )
    .fetch_optional(executor)
    .await?;

    Ok(record.map(|r| Tenant {
        id: r.id,
        name: r.name,
    }))
}

pub async fn update_tenant<'c, E>(
    executor: E,
    tenant_id: &str,
    tenant_name: &str,
) -> Result<Option<String>, TenantsDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        update app.tenants
        set name = $1
        where id = $2
        returning id
        "#,
        tenant_name,
        tenant_id
    )
    .fetch_optional(executor)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn delete_tenant<'c, E>(
    executor: E,
    tenant_id: &str,
) -> Result<Option<String>, TenantsDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        delete from app.tenants
        where id = $1
        returning id
        "#,
        tenant_id
    )
    .fetch_optional(executor)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn read_all_tenants<'c, E>(executor: E) -> Result<Vec<Tenant>, TenantsDbError>
where
    E: PgExecutor<'c>,
{
    let mut record = sqlx::query!(
        r#"
        select id, name
        from app.tenants
        "#,
    )
    .fetch_all(executor)
    .await?;

    Ok(record
        .drain(..)
        .map(|r| Tenant {
            id: r.id,
            name: r.name,
        })
        .collect())
}
