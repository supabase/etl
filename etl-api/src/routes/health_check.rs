use actix_web::{HttpResponse, Responder, get};

#[utoipa::path(
    summary = "API health status",
    description = "Returns 'ok' when the API is available and responding.",
    responses(
        (status = 200, description = "Health check passed; returns 'ok'.", body = String),
    ),
    tag = "Health",
)]
#[get("/health_check")]
pub async fn health_check() -> impl Responder {
    HttpResponse::Ok().body("ok")
}
