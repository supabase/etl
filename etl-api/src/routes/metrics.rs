use actix_web::{Responder, get, web};
use metrics_exporter_prometheus::PrometheusHandle;

#[utoipa::path(
    summary = "Get prometheus metrics",
    description = "Returns prometheus metrics collected since the last call to this endpoint.",
    responses(
        (status = 200, description = "Metrics returned successfully", body = String),
    ),
    tag = "Metrics"
)]
#[get("/metrics")]
pub(crate) async fn metrics(metrics_handle: web::ThinData<PrometheusHandle>) -> impl Responder {
    metrics_handle.render()
}
