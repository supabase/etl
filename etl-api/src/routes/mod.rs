use actix_web::HttpRequest;
use etl_config::shared::PgConnectionConfig;
use etl_postgres::replication::connect_to_source_database;
use serde::{Deserialize, Serialize};
use sqlx::PgPool;
use thiserror::Error;
use utoipa::ToSchema;

pub mod destinations;
pub mod destinations_pipelines;
pub mod health_check;
pub mod images;
pub mod pipelines;
pub mod sources;
pub mod tenants;
pub mod tenants_sources;

/// Minimum number of connections for the source Postgres connection pool.
const MIN_POOL_CONNECTIONS: u32 = 1;
/// Maximum number of connections for the source Postgres connection pool.
const MAX_POOL_CONNECTIONS: u32 = 1;

#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct ErrorMessage {
    #[schema(example = "an error occurred in the api")]
    pub error: String,
}

#[derive(Debug, Error)]
pub enum TenantIdError {
    #[error("The tenant id is missing in the request")]
    TenantIdMissing,

    #[error("The tenant id in the request is invalid")]
    TenantIdIllFormed,
}

fn extract_tenant_id(req: &HttpRequest) -> Result<&str, TenantIdError> {
    let headers = req.headers();
    let tenant_id = headers
        .get("tenant_id")
        .ok_or(TenantIdError::TenantIdMissing)?
        .to_str()
        .map_err(|_| TenantIdError::TenantIdIllFormed)?;

    Ok(tenant_id)
}

pub async fn connect_to_source_database_with_defaults(
    config: &PgConnectionConfig,
) -> Result<PgPool, sqlx::Error> {
    connect_to_source_database(config, MIN_POOL_CONNECTIONS, MAX_POOL_CONNECTIONS).await
}
