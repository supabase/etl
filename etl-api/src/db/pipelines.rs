use etl_config::shared::BatchConfig;
use serde::{Deserialize, Serialize};
use sqlx::{PgExecutor, PgTransaction};
use std::ops::DerefMut;
use thiserror::Error;
use utoipa::ToSchema;

use crate::db::replicators::{ReplicatorsDbError, create_replicator};
use crate::db::serde::{
    DbDeserializationError, DbSerializationError, deserialize_from_value, serialize,
};

/// Pipeline configuration used during replication. This struct's fields
/// should be kept in sync with [`OptionalPipelineConfig`]. If a new optional
/// field is added, it should also be included in the [`OptionalPipelineConfig::merge`]
/// implementation.
///
/// A separate struct was created because `publication_name` is not optional and
/// when updating config we do not want the user to pass publication
/// name.
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct PipelineConfig {
    #[schema(example = "my_publication")]
    pub publication_name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub batch: Option<BatchConfig>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub table_error_retry_delay_ms: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[schema(example = 4)]
    pub max_table_sync_workers: Option<u16>,
}

/// Has the same fields as [`PipelineConfig`] except from
/// the required fields. These two structs should be kept
/// in sync.
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct OptionalPipelineConfig {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub batch: Option<BatchConfig>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub table_error_retry_delay_ms: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_table_sync_workers: Option<u16>,
}

impl PipelineConfig {
    /// Merges an [`OptionalPipelineConfig`] into this one by overwriting the optional
    /// fields in self if they are set in the other config. There is currently no
    /// way to unset fields in self, but should be good enough for now.
    fn merge(&mut self, other: OptionalPipelineConfig) {
        if let Some(batch) = other.batch {
            self.batch = Some(batch);
        }

        if let Some(table_error_retry_delay_ms) = other.table_error_retry_delay_ms {
            self.table_error_retry_delay_ms = Some(table_error_retry_delay_ms);
        }

        if let Some(max_table_sync_workers) = other.max_table_sync_workers {
            self.max_table_sync_workers = Some(max_table_sync_workers);
        }
    }
}

pub struct Pipeline {
    pub id: i64,
    pub tenant_id: String,
    pub source_id: i64,
    pub source_name: String,
    pub destination_id: i64,
    pub destination_name: String,
    pub replicator_id: i64,
    pub config: PipelineConfig,
}

#[derive(Debug, Error)]
pub enum PipelinesDbError {
    #[error("Error while interacting with PostgreSQL for pipelines: {0}")]
    Database(#[from] sqlx::Error),

    #[error("Error while serializing pipeline config: {0}")]
    DbSerialization(#[from] DbSerializationError),

    #[error("Error while deserializing pipeline config: {0}")]
    DbDeserialization(#[from] DbDeserializationError),

    #[error(transparent)]
    ReplicatorsDb(#[from] ReplicatorsDbError),
}

pub async fn create_pipeline(
    txn: &mut PgTransaction<'_>,
    tenant_id: &str,
    source_id: i64,
    destination_id: i64,
    image_id: i64,
    config: PipelineConfig,
) -> Result<i64, PipelinesDbError> {
    let config = serialize(&config)?;

    let replicator_id = create_replicator(txn.deref_mut(), tenant_id, image_id).await?;
    let record = sqlx::query!(
        r#"
        insert into app.pipelines (tenant_id, source_id, destination_id, replicator_id, config)
        values ($1, $2, $3, $4, $5)
        returning id
        "#,
        tenant_id,
        source_id,
        destination_id,
        replicator_id,
        config
    )
    .fetch_one(txn.deref_mut())
    .await?;

    Ok(record.id)
}

pub async fn read_pipeline<'c, E>(
    executor: E,
    tenant_id: &str,
    pipeline_id: i64,
) -> Result<Option<Pipeline>, PipelinesDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        select p.id,
            p.tenant_id,
            source_id,
            s.name as source_name,
            destination_id,
            d.name as destination_name,
            replicator_id,
            p.config
        from app.pipelines p
        join app.sources s on p.source_id = s.id
        join app.destinations d on p.destination_id = d.id
        where p.tenant_id = $1 and p.id = $2
        "#,
        tenant_id,
        pipeline_id,
    )
    .fetch_optional(executor)
    .await?;

    let pipeline = match record {
        Some(record) => {
            let config = deserialize_from_value::<PipelineConfig>(record.config)?;

            let pipeline = Pipeline {
                id: record.id,
                tenant_id: record.tenant_id,
                source_id: record.source_id,
                source_name: record.source_name,
                destination_id: record.destination_id,
                destination_name: record.destination_name,
                replicator_id: record.replicator_id,
                config,
            };

            Some(pipeline)
        }
        None => None,
    };

    Ok(pipeline)
}

pub async fn update_pipeline<'c, E>(
    executor: E,
    tenant_id: &str,
    pipeline_id: i64,
    source_id: i64,
    destination_id: i64,
    config: &PipelineConfig,
) -> Result<Option<i64>, PipelinesDbError>
where
    E: PgExecutor<'c>,
{
    let pipeline_config = serialize(config)?;

    let record = sqlx::query!(
        r#"
        update app.pipelines
        set source_id = $1, destination_id = $2, config = $3
        where tenant_id = $4 and id = $5
        returning id
        "#,
        source_id,
        destination_id,
        pipeline_config,
        tenant_id,
        pipeline_id
    )
    .fetch_optional(executor)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn delete_pipeline<'c, E>(
    executor: E,
    tenant_id: &str,
    pipeline_id: i64,
) -> Result<Option<i64>, PipelinesDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        delete from app.pipelines
        where tenant_id = $1 and id = $2
        returning id
        "#,
        tenant_id,
        pipeline_id
    )
    .fetch_optional(executor)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn read_all_pipelines<'c, E>(
    executor: E,
    tenant_id: &str,
) -> Result<Vec<Pipeline>, PipelinesDbError>
where
    E: PgExecutor<'c>,
{
    let records = sqlx::query!(
        r#"
        select p.id,
            p.tenant_id,
            source_id,
            s.name as source_name,
            destination_id,
            d.name as destination_name,
            replicator_id,
            p.config
        from app.pipelines p
        join app.sources s on p.source_id = s.id
        join app.destinations d on p.destination_id = d.id
        where p.tenant_id = $1
        "#,
        tenant_id,
    )
    .fetch_all(executor)
    .await?;

    let mut pipelines = Vec::with_capacity(records.len());
    for record in records {
        let config = deserialize_from_value::<PipelineConfig>(record.config.clone())
            .expect("failed to deserialize pipeline config");

        pipelines.push(Pipeline {
            id: record.id,
            tenant_id: record.tenant_id,
            source_id: record.source_id,
            source_name: record.source_name,
            destination_id: record.destination_id,
            destination_name: record.destination_name,
            replicator_id: record.replicator_id,
            config,
        });
    }

    Ok(pipelines)
}

pub async fn update_pipeline_config(
    txn: &mut PgTransaction<'_>,
    tenant_id: &str,
    pipeline_id: i64,
    request_config: OptionalPipelineConfig,
) -> Result<Option<PipelineConfig>, PipelinesDbError> {
    // We use `select ... for update` to lock the pipeline row being updated
    // to avoid concurrent requests clobbering data from each other
    let record = sqlx::query!(
        r#"
        select p.id,
            p.config
        from app.pipelines p
        where p.tenant_id = $1 and p.id = $2
        for update
        "#,
        tenant_id,
        pipeline_id,
    )
    .fetch_optional(txn.deref_mut())
    .await?;

    match record {
        Some(record) => {
            let mut config_in_db = deserialize_from_value::<PipelineConfig>(record.config)?;
            config_in_db.merge(request_config);
            let updated_config = serialize(config_in_db)?;

            let record = sqlx::query!(
                r#"
                update app.pipelines
                set config = $1
                where tenant_id = $2 and id = $3
                returning config
                "#,
                updated_config,
                tenant_id,
                pipeline_id
            )
            .fetch_optional(txn.deref_mut())
            .await?;

            match record {
                Some(record) => {
                    let config = deserialize_from_value::<PipelineConfig>(record.config)?;
                    Ok(Some(config))
                }
                None => Ok(None),
            }
        }
        None => Ok(None),
    }
}

/// Helper function to check if an sqlx error is a duplicate pipeline constraint violation
pub fn is_duplicate_pipeline_error(err: &sqlx::Error) -> bool {
    match err {
        sqlx::Error::Database(db_err) => {
            // 23505 is PostgreSQL's unique constraint violation code
            // Check for our unique constraint name defined
            // in the migrations/20250605064229_add_unique_constraint_pipelines_source_destination.sql file
            db_err.code().as_deref() == Some("23505")
                && db_err.constraint() == Some("pipelines_tenant_source_destination_unique")
        }
        _ => false,
    }
}
