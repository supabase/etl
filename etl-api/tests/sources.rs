use etl_api::routes::pipelines::{CreatePipelineRequest, CreatePipelineResponse};
use etl_api::routes::sources::{
    CreateSourceRequest, CreateSourceResponse, ReadSourceResponse, ReadSourcesResponse,
    UpdateSourceRequest,
};
use etl_telemetry::tracing::init_test_tracing;
use reqwest::StatusCode;

use crate::support::mocks::create_default_image;
use crate::support::mocks::destinations::create_destination;
use crate::support::mocks::pipelines::new_pipeline_config;
use crate::support::mocks::sources::{
    create_source, create_source_with_config, new_name, new_source_config, updated_name,
    updated_source_config,
};
use crate::support::mocks::tenants::create_tenant;
use crate::support::test_app::spawn_test_app;

mod support;

#[tokio::test(flavor = "multi_thread")]
async fn source_can_be_created() {
    init_test_tracing();
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;

    // Act
    let source = CreateSourceRequest {
        name: new_name(),
        config: new_source_config(),
    };
    let response = app.create_source(tenant_id, &source).await;

    // Assert
    assert!(response.status().is_success());
    let response: CreateSourceResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    assert_eq!(response.id, 1);
}

#[tokio::test(flavor = "multi_thread")]
async fn an_existing_source_can_be_read() {
    init_test_tracing();
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;

    let source = CreateSourceRequest {
        name: new_name(),
        config: new_source_config(),
    };
    let response = app.create_source(tenant_id, &source).await;
    let response: CreateSourceResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    let source_id = response.id;

    // Act
    let response = app.read_source(tenant_id, source_id).await;

    // Assert
    assert!(response.status().is_success());
    let response: ReadSourceResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    assert_eq!(response.id, source_id);
    assert_eq!(&response.tenant_id, tenant_id);
    assert_eq!(response.name, source.name);
    insta::assert_debug_snapshot!(response.config);
}

#[tokio::test(flavor = "multi_thread")]
async fn non_existing_source_cannot_be_read() {
    init_test_tracing();
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;

    // Act
    let response = app.read_source(tenant_id, 42).await;

    // Assert
    assert_eq!(response.status(), StatusCode::NOT_FOUND);
}

#[tokio::test(flavor = "multi_thread")]
async fn an_existing_source_can_be_updated() {
    init_test_tracing();
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;

    let source = CreateSourceRequest {
        name: new_name(),
        config: new_source_config(),
    };
    let response = app.create_source(tenant_id, &source).await;
    let response: CreateSourceResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    let source_id = response.id;

    // Act
    let updated_config = UpdateSourceRequest {
        name: updated_name(),
        config: updated_source_config(),
    };
    let response = app
        .update_source(tenant_id, source_id, &updated_config)
        .await;

    // Assert
    assert!(response.status().is_success());
    let response = app.read_source(tenant_id, source_id).await;
    let response: ReadSourceResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    assert_eq!(response.id, source_id);
    assert_eq!(&response.tenant_id, tenant_id);
    assert_eq!(response.name, updated_config.name);
    insta::assert_debug_snapshot!(response.config);
}

#[tokio::test(flavor = "multi_thread")]
async fn non_existing_source_cannot_be_updated() {
    init_test_tracing();
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;

    // Act
    let updated_config = UpdateSourceRequest {
        name: updated_name(),
        config: updated_source_config(),
    };
    let response = app.update_source(tenant_id, 42, &updated_config).await;

    // Assert
    assert_eq!(response.status(), StatusCode::NOT_FOUND);
}

#[tokio::test(flavor = "multi_thread")]
async fn an_existing_source_can_be_deleted() {
    init_test_tracing();
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;

    let source = CreateSourceRequest {
        name: new_name(),
        config: new_source_config(),
    };
    let response = app.create_source(tenant_id, &source).await;
    let response: CreateSourceResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    let source_id = response.id;

    // Act
    let response = app.delete_source(tenant_id, source_id).await;

    // Assert
    assert!(response.status().is_success());
    let response = app.read_source(tenant_id, source_id).await;
    assert_eq!(response.status(), StatusCode::NOT_FOUND);
}

#[tokio::test(flavor = "multi_thread")]
async fn non_existing_source_cannot_be_deleted() {
    init_test_tracing();
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;

    // Act
    let response = app.delete_source(tenant_id, 42).await;

    // Assert
    assert_eq!(response.status(), StatusCode::NOT_FOUND);
}

#[tokio::test(flavor = "multi_thread")]
async fn all_sources_can_be_read() {
    init_test_tracing();
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;
    let source1_id =
        create_source_with_config(&app, tenant_id, new_name(), new_source_config()).await;
    let source2_id =
        create_source_with_config(&app, tenant_id, updated_name(), updated_source_config()).await;

    // Act
    let response = app.read_all_sources(tenant_id).await;

    // Assert
    assert!(response.status().is_success());
    let response: ReadSourcesResponse = response
        .json()
        .await
        .expect("failed to deserialize response");
    for source in response.sources {
        if source.id == source1_id {
            let name = new_name();
            assert_eq!(&source.tenant_id, tenant_id);
            assert_eq!(source.name, name);
            insta::assert_debug_snapshot!(source.config);
        } else if source.id == source2_id {
            let name = updated_name();
            assert_eq!(&source.tenant_id, tenant_id);
            assert_eq!(source.name, name);
            insta::assert_debug_snapshot!(source.config);
        }
    }
}

#[tokio::test(flavor = "multi_thread")]
async fn source_with_active_pipeline_cannot_be_deleted() {
    init_test_tracing();
    // Arrange
    let app = spawn_test_app().await;
    let tenant_id = &create_tenant(&app).await;
    create_default_image(&app).await;

    // Create source and destination
    let source_id = create_source(&app, tenant_id).await;
    let destination_id = create_destination(&app, tenant_id).await;

    // Create a pipeline that uses this source
    let pipeline = CreatePipelineRequest {
        source_id,
        destination_id,
        config: new_pipeline_config(),
    };
    let pipeline_response = app.create_pipeline(tenant_id, &pipeline).await;
    assert!(pipeline_response.status().is_success());
    let pipeline_response: CreatePipelineResponse = pipeline_response
        .json()
        .await
        .expect("failed to deserialize response");
    let _pipeline_id = pipeline_response.id;

    // Act - Try to delete the source
    let response = app.delete_source(tenant_id, source_id).await;

    // Assert - Should fail due to foreign key constraint
    assert_eq!(response.status(), StatusCode::INTERNAL_SERVER_ERROR);

    // Verify source still exists
    let source_response = app.read_source(tenant_id, source_id).await;
    assert!(source_response.status().is_success());
}
