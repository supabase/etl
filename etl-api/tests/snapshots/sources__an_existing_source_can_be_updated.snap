---
source: etl-api/tests/sources.rs
expression: response.config
---
StrippedApiSourceConfig {
    host: "example.com",
    port: 2345,
    name: "sergtsop",
    username: "sergtsop",
}
